drivers/eeprom/include/eeprom_driver.h

```diff
@@ -0,0 +1,128 @@
+     * @brief   Initialize the data structures for the EEPROM device
+     *
+     * @param[in,out]   handle  Device handle to initialize
+     * @param[in]       params  Initialization parameters of the device
+     *
+     * @retval  0       Failed to initialize the device
+     * @return          The size of the device in bytes
+     */
+    eeprom_off_t (*init)(void *handle, const void *params);
+    /**
+     * @brief   Read data from the EEPROM device
+     *
+     * @param[in,out]   handle  Device handle of the EEPROM to read from
+     * @param[in]       pos     Position to start to read from (in bytes)
+     * @param[out]      dest    Buffer to write the data to
+     * @param[in]       len     Number of bytes to read
+     *
+     * @retval  0       Success
+     * @retval  -EIO    I/O error occurred while reading
+     */
+    int (*read)(void *handle, eeprom_off_t pos, void *dest, size_t len);
+    /**
+     * @brief   Write data from the EEPROM device
+     *
+     * @param[in,out]   handle  Device handle of the EEPROM to write to
+     * @param[in]       pos     Position to start to write to (in bytes)
+     * @param[in]       data    Buffer to write to the EEPROM device
+     * @param[in,out]   len     Number of bytes to write
+     *
+     * @retval  0       Success
+     * @retval  -EIO    I/O error occurred while writen
+     *
+     * If `-EFAULT` has been returned, the content of the EEPROM device is
+     * unchanged. If `0` is returned, all data has been written successfully
+     * to the device. If `-EIO` is returned, the contents in `[pos; pos + len[`
+     * of the device are undefined, but the remaining content of the device
+     * is unchanged.
+     */
+    int (*write)(void *handle, eeprom_off_t pos, const void *data, size_t len);
+    /**
+     * @brief   Erase the whole EEPROM
+     *
+     * @param[in,out]   handle  Device handle of the EEPROM to erase
+     *
+     * @retval  0       Success
+     * @retval  -EIO    I/O error occurred during erase
+     *
+     * If `0` is returned a subsequent read for any valid address must return
+     * @ref eeprom_driver_t::clear_byte. If `-EIO` is returned, the whole
+     * content of the EEPROM is undefined.
+     *
+     * This function is optional to implement by the driver. If the function
+     * pointer is `NULL`, a fallback implementation that calls @ref eeprom_set
+     * is used instead. This will fill the EEPROM with
+     * @ref eeprom_driver_t::clear_byte.
+     */
+    int (*erase)(void *handle);
+    /**
+     * @brief   The EEPROM is filled with this byte after being erased
+     */
+    uint8_t clear_byte;
+} eeprom_driver_t;
+
+/**
+ * @brief   The current state of an EEPROM device
+ */
+typedef struct {
+    const eeprom_driver_t *driver;  /**< The driver of the EEPROM device */
+    const void *params;             /**< The initialization parameters */
+    void *handle;                   /**< The handle to pass to the driver */
+} eeprom_dev_t;
```

Comment:
@aabadie: I think if we drop the `params` in the init callback, it will no longer be possible to add multiple external EEPROM devices that are using the same driver, without some dark magic. Assume the init function of an external driver is only called with the device handle to initialize. With what parameters should it initialize that device handle?

If there is only one device of that kind, it is clear. But RIOT policy is to allow multiple devices of a kind. I don't see how this could work.